### PR TITLE
G-API: Wrap GStreamerSource

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration
```
build_image:Linux AVX2=ubuntu:20.04

force_builders=Custom,Custom Win,Custom Mac,Linux AVX2
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.4.1:20.04
build_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp [hunk 1]
```diff
@@ -19,12 +19,12 @@ namespace gapi {
...
 namespace gst {
 
-class GAPI_EXPORTS GStreamerPipeline
+class GAPI_EXPORTS_W GStreamerPipeline
 {
 public:
     class Priv;
 
-    explicit GStreamerPipeline(const std::string& pipeline);
+    CV_WRAP explicit GStreamerPipeline(const std::string& pipeline);
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:27]: `CV_WRAP` -> `GAPI_WRAP`
Not sure that it's needed to wrap this `ctor` since `GStreamerPipeline` isn't supposed to be used, instead of this we've wrapped `make_gst_src`

Ok, let it be

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp [hunk 2]
```diff
@@ -40,6 +40,15 @@ class GAPI_EXPORTS GStreamerPipeline
 
 using GStreamerPipeline = gst::GStreamerPipeline;
 
+// NB: Function for using from python
+GAPI_EXPORTS_W cv::Ptr<IStreamSource>
+inline get_streaming_source(cv::Ptr<GStreamerPipeline>& pipeline,
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:45]: Why it's needed?

File: modules/gapi/misc/python/package/gapi/__init__.py [hunk 3]
```diff
@@ -297,3 +297,5 @@ def kernel_with_params(cls):
 cv.gapi.wip.draw.Poly = cv.gapi_wip_draw_Poly
 
 cv.gapi.streaming.queue_capacity = cv.gapi_streaming_queue_capacity
+
+cv.gapi.wip.gst.GStreamerPipeline = cv.gapi.wip.GStreamerPipeline = cv.gapi_wip_gst_GStreamerPipeline
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/package/gapi/__init__.py:301]: ???

File: modules/gapi/misc/python/pyopencv_gapi.hpp [hunk 4]
```diff
@@ -1074,6 +1075,5 @@ bool pyopencv_to(PyObject* obj, cv::GProtoOutputArgs& value, const ArgInfo& info
   {"kernels", CV_PY_FN_WITH_KW(pyopencv_cv_gapi_kernels), "kernels(...) -> GKernelPackage"}, \
   {"__op", CV_PY_FN_WITH_KW(pyopencv_cv_gapi_op), "__op(...) -> retval\n"},
 
-
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/pyopencv_gapi.hpp:1077]: Unnecessary change

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 5]
```diff
@@ -229,7 +229,6 @@ def test_video_good_features_to_track(self):
 
 
         def test_gapi_streaming_meta(self):
-            ksize = 3
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:232]: Is it necessary?

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 6]
```diff
@@ -350,6 +349,67 @@ def test_compile_streaming_meta_and_args(self):
...
+            ccomp.setSource(cv.gin(source))
+            ccomp.start()
+
+            has_frame, output = ccomp.pull()
+            while has_frame:
+                self.assertTrue(output.size != 0)
+                has_frame, output = ccomp.pull()
+
+
+        def test_gstreamer_pipeline(self):
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:378]: What's the difference between: `test_gstreamer_source` and `test_gstreamer_pipeline` now sure that test names reflect it

I'd name test like this: `test_gst_two_inputs` or something like this

Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:378]: OK, makes sense
It is kind of testing `GStreamerPipeline` class wrapped functionality (with testing multiple inputs, too) - that's why it was named after it

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 5 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:232]: This `ksize` isn't used in this particular test, I've just noticed that and decided to clear that

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 1 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp]: Actually GStreamerPipeline is as essential as GStreamerSource because you cannot create multi-source Gst-pipelines without it

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 2 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:45]: I did try to wrap just the method `GStreamerPipeline::getStreamingSource()` - but didn't succeed. The problem is in the returning value

I can try to find better solution (probably, wrap the method somehow)

File: modules/gapi/misc/python/package/gapi/__init__.py (hunk 3 above)
Comment (OrestChura) [reply modules/gapi/misc/python/package/gapi/__init__.py:301]: Don't know but it's supported by ocv-version: https://github.com/opencv/opencv/blob/4.x/modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamersource.hpp#L83

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 2 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp]: rewritten wrapping: now it is a class method, not a separate function.
More convenient to use:
```
pipeline = """some ! gst ! pipeline ! appsink name=sink1
              some ! gst ! pipeline ! appsink name=sink2"""
pp = cv.gapi.wip.GStreamerPipeline(pipeline)
src1 = pp.get_streaming_source("sink1")   # instead of src1 = cv.gapi.wip.get_streaming_source(pp, "sink1")
src2 = pp.get_streaming_source("sink2")
...
ccomp = c.compileStreaming()
ccomp.setSource(cv.gin(src1, src2))
...
```

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp [hunk 7]
```diff
@@ -19,15 +19,19 @@ namespace gapi {
...
 {
 public:
     class Priv;
 
-    explicit GStreamerPipeline(const std::string& pipeline);
+    GAPI_WRAP explicit GStreamerPipeline(const std::string& pipeline);
     IStreamSource::Ptr getStreamingSource(const std::string& appsinkName,
                                           const GStreamerSource::OutputType outputType =
                                               GStreamerSource::OutputType::MAT);
+    // NB: Function for using from python
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:31]: In fact it's  class `method`, I wouldn't put such obvious comment, since you mark it as `GAPI_WRAP`

File: modules/gapi/misc/python/package/gapi/__init__.py (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/misc/python/package/gapi/__init__.py:301]: I meant double assignments

Comment (OrestChura) [reply modules/gapi/misc/python/package/gapi/__init__.py]: OK, removed `wip.gst.` assignment

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 7 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp]: OK, removed

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 8]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+                return cv.gapi.wip.GStreamerPipeline(gstpipeline)
+            except cv.error as e:
+                if str(e).find('Built without GStreamer support!') == -1:
+                    raise e
+                else:
+                    raise unittest.SkipTest(str(e))
+            except SystemError as e:
+                raise unittest.SkipTest(str(e) + ", casued by " + str(e.__cause__))
+
+        def test_gst_multiple_sources(self):
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:457]: What's the difference between `multiple_sources` and `multisource` bellow? 

Do we really need such tests with `self.assertTrue(output.size != 0)` if we have accuracy tests?

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 9]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+            cap2 = self.open_VideoCapture_gstreamer(gstpipeline2)
+
+            # Assert
+            max_num_frames = 10
+            for _ in range(max_num_frames):
+                has_expected1, expected1 = cap1.read()
+                has_expected2, expected2 = cap2.read()
+                has_actual, (actual1, actual2) = ccomp.pull()
+
+                self.assertEqual(has_expected1, has_expected2)
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:531]: `self.assertTrue(has_expected1 and has_expected2)`
`self.assertTrue(has_actual1 and has_actual2)`

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 10]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+                has_expected = has_expected1 and has_expected2
+                self.assertEqual(has_expected, has_actual)
+
+                if not has_expected:
+                    break
+
+                self.assertEqual(0.0, cv.norm(convertNV12p2BGR(expected1), actual1, cv.NORM_INF))
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:538]: Why `exepcted` and `actual` have different color formats ?

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp [hunk 11]
```diff
@@ -19,15 +19,18 @@ namespace gapi {
...
 {
 public:
     class Priv;
 
-    explicit GStreamerPipeline(const std::string& pipeline);
+    GAPI_WRAP explicit GStreamerPipeline(const std::string& pipeline);
     IStreamSource::Ptr getStreamingSource(const std::string& appsinkName,
                                           const GStreamerSource::OutputType outputType =
                                               GStreamerSource::OutputType::MAT);
+    GAPI_WRAP cv::Ptr<cv::gapi::wip::IStreamSource> get_streaming_source(
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:31]: Why don't you wrap `getStreamingSource` instead of implementing the new one?

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 12]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+        def get_gst_source(self, gstpipeline):
+            # NB: Skip test in case gstreamer isn't available.
+            try:
+                return cv.gapi.wip.make_gst_src(gstpipeline)
+            except cv.error as e:
+                if str(e).find('Built without GStreamer support!') == -1:
+                    raise e
+                else:
+                    raise unittest.SkipTest(str(e))
+
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:371]: Two lines

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 13]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+                if not has_expected:
+                    break
+
+                self.assertEqual(0.0, cv.norm(convertNV12p2BGR(expected), actual, cv.NORM_INF))
+
+
+        def get_gst_pipeline(self, gstpipeline):
+            # NB: Skip test in case gstreamer isn't available.
+            try:
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/test/test_gapi_streaming.py:447]: Do we really need such complicated exception check?

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 8 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:457]: For the naming - agree, tests need to be named consistently, and there are no diffrerence between "multiple_sources" and "multisource"; will rename them to `test_gst_multiple_sources` and `test_gst_multiple_sources_accuracy`

the `accuracy` tests use real images, and I think it is useful to have a different test not depending on real images which just checks if gapi gstreamer pipeline works somehow. I think we should keep both tests

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 9 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:531]: Cannot agree - `has_expected` and `has_actual` can turn `False`. It was intentional that I check equality of states of two sources.

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 10 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:538]: Long story short, `cv::VideoCapture(..., cv.CAP_GSTREAMER)`'s output is plain/combined-in-one-Mat NV12; our source outputs BGR.

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 11 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:31]: the new method has different output type;

I didn't manage to make `IStreamSource::Ptr` working in Python

also, as I understand, this is why we have `make_capture_src` and `make_gst_src` "Python-only" functions instead of just wrapping already existing `make_src`, besides it is a template

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 13 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:447]: I don't know.. I thought we should be sure we skip a test only because it can be performed - in this case, only if OpenCV is built without GStreamer

If you say, I will not check that, but I see it as a useful check

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 14]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+        def open_VideoCapture_gstreamer(self, gstpipeline):
+            try:
+                cap = cv.VideoCapture(gstpipeline, cv.CAP_GSTREAMER)
+            except Exception as e:
+                raise unittest.SkipTest("Backend GSTREAMER can't open the video; " +
+                                        "cause: " + str(e))
+            if not cap.isOpened():
+                raise unittest.SkipTest("Backend GSTREAMER can't open the video")
+            return cap
+
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:407]: OK done

File: modules/gapi/misc/python/test/test_gapi_streaming.py [hunk 15]
```diff
@@ -350,6 +359,186 @@ def test_compile_streaming_meta_and_args(self):
...
+            try:
+                return cv.gapi.wip.GStreamerPipeline(gstpipeline)
+            except cv.error as e:
+                if str(e).find('Built without GStreamer support!') == -1:
+                    raise e
+                else:
+                    raise unittest.SkipTest(str(e))
+            except SystemError as e:
+                raise unittest.SkipTest(str(e) + ", casued by " + str(e.__cause__))
+
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:458]: done

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 8 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:460]: renamed test

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 12 above)
Comment (OrestChura) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:371]: done

File: modules/gapi/misc/python/test/test_gapi_streaming.py (hunk 13 above)
Comment (TolyaTalamanov) [reply modules/gapi/misc/python/test/test_gapi_streaming.py:449]: Ok

File: modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp (hunk 2 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/streaming/gstreamer/gstreamerpipeline.hpp:47]: I had problems with wrapping the method as it's compllicated to wrap the returning class `IStreamSource::Ptr = std::shared_ptr<IStreamSource>`

set to have a separate functions for now

left a comment about it

**Review by TolyaTalamanov — COMMENTED** [review]
LGTM 👍

**Review by dmatveev — APPROVED** [review]
